app/scripts/lib/rpc-method-middleware/handlers/switch-ethereum-chain.js

```diff
@@ -1,168 +1,96 @@
 
 const switchEthereumChain = {
   methodNames: [MESSAGE_TYPE.SWITCH_ETHEREUM_CHAIN],
   implementation: switchEthereumChainHandler,
   hookNames: {
-    getCurrentChainId: true,
     findNetworkConfigurationBy: true,
-    findNetworkClientIdByChainId: true,
-    setNetworkClientIdForDomain: true,
-    setProviderType: true,
     setActiveNetwork: true,
+    getCaveat: true,
+    requestPermittedChainsPermission: true,
+    getCurrentChainIdForDomain: true,
     requestUserApproval: true,
-    getNetworkConfigurations: true,
-    getProviderConfig: true,
-    hasPermissions: true,
+    getChainPermissionsFeatureFlag: true,
   },
 };
 
 export default switchEthereumChain;
 
-function findExistingNetwork(chainId, findNetworkConfigurationBy) {
-  if (
-    Object.values(BUILT_IN_INFURA_NETWORKS)
-      .map(({ chainId: id }) => id)
-      .includes(chainId)
-  ) {
-    return {
-      chainId,
-      ticker: CURRENCY_SYMBOLS.ETH,
-      nickname: NETWORK_TO_NAME_MAP[chainId],
-      rpcUrl: CHAIN_ID_TO_RPC_URL_MAP[chainId],
-      type: CHAIN_ID_TO_TYPE_MAP[chainId],
-    };
-  }
-
-  return findNetworkConfigurationBy({ chainId });
-}
-
 async function switchEthereumChainHandler(
   req,
   res,
   _next,
   end,
   {
-    getCurrentChainId,
     findNetworkConfigurationBy,
-    findNetworkClientIdByChainId,
-    setNetworkClientIdForDomain,
-    setProviderType,
     setActiveNetwork,
+    requestPermittedChainsPermission,
+    getCaveat,
+    getCurrentChainIdForDomain,
     requestUserApproval,
-    getProviderConfig,
-    hasPermissions,
+    getChainPermissionsFeatureFlag,
   },
 ) {
-  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected single, object parameter. Received:\n${JSON.stringify(
-          req.params,
-        )}`,
-      }),
-    );
+  let chainId;
+  try {
+    chainId = validateSwitchEthereumChainParams(req, end);
+  } catch (error) {
+    return end(error);
   }
 
   const { origin } = req;
-
-  const { chainId } = req.params[0];
-
-  const otherKeys = Object.keys(omit(req.params[0], ['chainId']));
-
-  if (otherKeys.length > 0) {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
-      }),
-    );
+  const currentChainIdForOrigin = getCurrentChainIdForDomain(origin);
+  if (currentChainIdForOrigin === chainId) {
+    res.result = null;
+    return end();
   }
 
-  const _chainId = typeof chainId === 'string' && chainId.toLowerCase();
+  const networkConfigurationForRequestedChainId = findExistingNetwork(
+    chainId,
+    findNetworkConfigurationBy,
+  );
 
-  if (!isPrefixedFormattedHexString(_chainId)) {
-    return end(
-      ethErrors.rpc.invalidParams({
-        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
-      }),
-    );
-  }
+  const networkClientIdToSwitchTo =
```

Comment:
Heads up that this will change following https://github.com/MetaMask/core/pull/4286, and you'll probably want something like:

``` typescript
const networkConfigurationForRequestedChainId =
  getNetworkConfigurationByChainId(chainId);

const networkClientIdToSwitchTo = networkConfigurationForRequestedChainId 
 
 .rpcEndpointsByUrl[networkConfigurationForRequestedChainId.defaultRpcEndpointUrl]
  .networkClientId;
```